Description: Adds the fields "Langlist" and "LCMess" to dmrc, which now can hold the LANGUAGE respective LC_MESSAGES values.
Bug-Ubuntu: https://bugs.launchpad.net/bugs/553162
Forwarded: not-needed
Author: Gunnar Hjalmarsson <ubuntu@gunnar.cc>

diff -Nur -x '*.orig' -x '*~' gdm-3.0.0//daemon/gdm-session-settings.c gdm-3.0.0.new//daemon/gdm-session-settings.c
--- gdm-3.0.0//daemon/gdm-session-settings.c	2011-06-05 23:18:55.000000000 +0200
+++ gdm-3.0.0.new//daemon/gdm-session-settings.c	2011-06-05 23:19:55.847165226 +0200
@@ -40,6 +40,8 @@
         ActUser *user;
         char *session_name;
         char *language_name;
+        char *langlist_name;
+        char *lcmess_name;
         gboolean failsafe;
 };
 
@@ -60,6 +62,8 @@
         PROP_0 = 0,
         PROP_SESSION_NAME,
         PROP_LANGUAGE_NAME,
+        PROP_LANGLIST_NAME,
+        PROP_LCMESS_NAME,
         PROP_IS_FAILSAFE,
         PROP_IS_LOADED
 };
@@ -101,6 +105,18 @@
                                        G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY);
         g_object_class_install_property (object_class, PROP_LANGUAGE_NAME, param_spec);
 
+        param_spec = g_param_spec_string ("langlist-name", "Language List Name",
+                                        "The name of the LANGUAGE list",
+                                        NULL,
+                                       G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY);
+        g_object_class_install_property (object_class, PROP_LANGLIST_NAME, param_spec);
+
+        param_spec = g_param_spec_string ("lcmess-name", "Messages Locale Name",
+                                        "The name of LC_MESSAGES",
+                                        NULL,
+                                       G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY);
+        g_object_class_install_property (object_class, PROP_LCMESS_NAME, param_spec);
+
         param_spec = g_param_spec_string ("is-failsafe", "Failsafe Session",
                                         "Whether the session is a failsafe one",
                                         NULL,
@@ -137,6 +153,8 @@
 
         g_free (settings->priv->session_name);
         g_free (settings->priv->language_name);
+        g_free (settings->priv->langlist_name);
+        g_free (settings->priv->lcmess_name);
 
         parent_class = G_OBJECT_CLASS (gdm_session_settings_parent_class);
 
@@ -159,6 +177,32 @@
 }
 
 void
+gdm_session_settings_set_langlist_name (GdmSessionSettings *settings,
+                                       const char         *langlist_name)
+{
+        g_return_if_fail (GDM_IS_SESSION_SETTINGS (settings));
+
+        if (settings->priv->langlist_name == NULL ||
+            strcmp (settings->priv->langlist_name, langlist_name) != 0) {
+                settings->priv->langlist_name = g_strdup (langlist_name);
+                g_object_notify (G_OBJECT (settings), "langlist-name");
+        }
+}
+
+void
+gdm_session_settings_set_lcmess_name (GdmSessionSettings *settings,
+                                       const char         *lcmess_name)
+{
+        g_return_if_fail (GDM_IS_SESSION_SETTINGS (settings));
+
+        if (settings->priv->lcmess_name == NULL ||
+            strcmp (settings->priv->lcmess_name, lcmess_name) != 0) {
+                settings->priv->lcmess_name = g_strdup (lcmess_name);
+                g_object_notify (G_OBJECT (settings), "lcmess-name");
+        }
+}
+
+void
 gdm_session_settings_set_session_name (GdmSessionSettings *settings,
                                        const char         *session_name)
 {
@@ -191,6 +235,20 @@
 }
 
 char *
+gdm_session_settings_get_langlist_name (GdmSessionSettings *settings)
+{
+        g_return_val_if_fail (GDM_IS_SESSION_SETTINGS (settings), NULL);
+        return g_strdup (settings->priv->langlist_name);
+}
+
+char *
+gdm_session_settings_get_lcmess_name (GdmSessionSettings *settings)
+{
+        g_return_val_if_fail (GDM_IS_SESSION_SETTINGS (settings), NULL);
+        return g_strdup (settings->priv->lcmess_name);
+}
+
+char *
 gdm_session_settings_get_session_name (GdmSessionSettings *settings)
 {
         g_return_val_if_fail (GDM_IS_SESSION_SETTINGS (settings), NULL);
@@ -219,6 +277,14 @@
                         gdm_session_settings_set_language_name (settings, g_value_get_string (value));
                 break;
 
+                case PROP_LANGLIST_NAME:
+                        gdm_session_settings_set_langlist_name (settings, g_value_get_string (value));
+                break;
+
+                case PROP_LCMESS_NAME:
+                        gdm_session_settings_set_lcmess_name (settings, g_value_get_string (value));
+                break;
+
                 case PROP_SESSION_NAME:
                         gdm_session_settings_set_session_name (settings, g_value_get_string (value));
                 break;
@@ -251,6 +317,14 @@
                         g_value_set_string (value, settings->priv->language_name);
                 break;
 
+                case PROP_LANGLIST_NAME:
+                        g_value_set_string (value, settings->priv->langlist_name);
+                break;
+
+                case PROP_LCMESS_NAME:
+                        g_value_set_string (value, settings->priv->lcmess_name);
+                break;
+
                 case PROP_IS_FAILSAFE:
                         g_value_set_boolean (value, settings->priv->failsafe);
                 break;
@@ -290,6 +364,8 @@
 {
         const char *session_name;
         const char *language_name;
+        const char *langlist_name;
+        const char *lcmess_name;
 
         if (!act_user_is_loaded (settings->priv->user)) {
                 g_warning ("GdmSessionSettings: trying to load user settings from unloaded user");
@@ -310,6 +386,18 @@
                 gdm_session_settings_set_language_name (settings, language_name);
         }
 
+        langlist_name = act_user_get_langlist (settings->priv->user);
+        g_debug ("GdmSessionSettings: saved langlist is %s", langlist_name);
+        if (langlist_name != NULL) {
+                gdm_session_settings_set_langlist_name (settings, langlist_name);
+        }
+
+        lcmess_name = act_user_get_lcmess (settings->priv->user);
+        g_debug ("GdmSessionSettings: saved lcmess is %s", lcmess_name);
+        if (lcmess_name != NULL) {
+                gdm_session_settings_set_lcmess_name (settings, lcmess_name);
+        }
+
         gdm_session_settings_set_is_failsafe (settings, FALSE);
 
         g_object_notify (G_OBJECT (settings), "is-loaded");
@@ -396,6 +484,15 @@
         if (settings->priv->language_name != NULL) {
                 act_user_set_language (user, settings->priv->language_name);
         }
+
+        if (settings->priv->langlist_name != NULL) {
+                act_user_set_langlist (user, settings->priv->langlist_name);
+        }
+
+        if (settings->priv->lcmess_name != NULL) {
+                act_user_set_lcmess (user, settings->priv->lcmess_name);
+        }
+
         g_object_unref (user);
 
         return TRUE;
diff -Nur -x '*.orig' -x '*~' gdm-3.0.0//daemon/gdm-session-settings.h gdm-3.0.0.new//daemon/gdm-session-settings.h
--- gdm-3.0.0//daemon/gdm-session-settings.h	2011-06-05 23:18:55.000000000 +0200
+++ gdm-3.0.0.new//daemon/gdm-session-settings.h	2011-06-05 23:20:57.417255523 +0200
@@ -59,10 +59,16 @@
                                                              const char          *username);
 gboolean            gdm_session_settings_is_loaded          (GdmSessionSettings  *settings);
 char               *gdm_session_settings_get_language_name  (GdmSessionSettings *settings);
+char               *gdm_session_settings_get_langlist_name  (GdmSessionSettings *settings);
+char               *gdm_session_settings_get_lcmess_name    (GdmSessionSettings *settings);
 char               *gdm_session_settings_get_session_name   (GdmSessionSettings *settings);
 gboolean            gdm_session_settings_get_is_failsafe    (GdmSessionSettings *settings);
 void                gdm_session_settings_set_language_name  (GdmSessionSettings *settings,
                                                              const char         *language_name);
+void                gdm_session_settings_set_langlist_name  (GdmSessionSettings *settings,
+                                                             const char         *langlist_name);
+void                gdm_session_settings_set_lcmess_name    (GdmSessionSettings *settings,
+                                                             const char         *lcmess_name);
 void                gdm_session_settings_set_session_name   (GdmSessionSettings *settings,
                                                              const char         *session_name);
 void                gdm_session_settings_set_is_failsafe    (GdmSessionSettings *settings,
